src/main/java/com/google/sps/servlets/ProjectServlet.java

```diff
@@ -0,0 +1,329 @@
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.CompositeFilter;
+import com.google.appengine.api.datastore.Query.CompositeFilterOperator;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.appengine.api.users.User;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import java.io.IOException;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.LinkedHashSet;
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/projects")
+public class ProjectServlet extends HttpServlet {
+
+  private static final String ASCENDING_SORT = "asc";
+  private static final String DESCENDING_SORT = "dsc";
+  private static final String PRIVATE = "private";
+  private static final String PUBLIC = "public";
+
+  /**
+   * Determines if the given request parameter is empty
+   * @param     {String}    param   request parameter
+   * @return    {Boolean}
+   */
+  private Boolean isEmptyParameter(String param) {
+    return param == null || param.isEmpty();
+  }
+
+  /**
+   * Removes duplicate values with a hash set
+   * @param     {ArrayList<String>} al   all values
+   * @return    {List<String>}
+   */
+  List<String> withDuplicatesRemoved(ArrayList<String> al) {
+    LinkedHashSet<String> lhs = new LinkedHashSet<String>(al);
+    return new ArrayList<String>(lhs);
+  }
+
+  /**
+   * Parses comma-separated list of emails into array
+   * @param     {String}            emails   unseparated emails
+   * @return    {ArrayList<String>}
+   */
+  ArrayList<String> parseEmails(String emails) {
+    return new ArrayList(
+        Arrays.asList(emails.toLowerCase().split("\\s*,\\s*")));
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response)
+      throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // Must be logged in
+    if (!userService.isUserLoggedIn()) {
+      response.sendRedirect("/");
+      return;
+    }
+
+    // Mode is a required parameter
+    String mode = request.getParameter("mode");
+    if (isEmptyParameter(mode) || (!mode.toLowerCase().equals("create") &&
```

Comment:
Are create and update the modes for saving a new project and updating an old one? just wanted to make sure I understand.

Reply:
Yes